### PR TITLE
Remove OptionValue and OptionType code

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Deletes all code related to `Spree::OptionValue` and `Spree::OptionType` objects :fire:

Part of this feature is from old Spree code, and part of it was hacked on and modified extensively in the very early days of OFN. It causes innumerable problems at every level of the app, primarily in terms of _performance_ but also at the level of the datamodel, making it really difficult in some cases to work with data in ways which should not be difficult.

here-be-dragons980213

The "option types" here are the three unit_value options, `weight`, `volume`, and `items`, and their "option values" are their displayed output for those types, like `1 loaf` or `125g bag` etc. 

Due to the way "option values" and "option types" are stored in the database; in order to get a variant's name or a line item's name we have to run a crazy process that involves 8 different classes and queries across several different tables. At scale in areas like reports where you might want to find out the name of 2000 line items at once, this can often generate staggering levels of complexity and vast amounts of additional queries which otherwise wouldn't be necessary just to find out that some carrots should be referred to as `1 bunch`. Part of the issue is that these values are never persisted directly in a database table, so they have to be calculated (by a crazy process) every time they are asked for instead of being directly available in a column.

Sooooo... it turns out we can actually just rip out the entire thing and delete it and the app will be radically simpler and faster. Everything should work exactly as before, but with significantly less database hits in many parts of the app. Also previous issues like Zapier or Matabase not being able to easily find what an item is called will be resolved, it'll be trivial to query those values in direct and simple SQL queries.

#### What should we test?

Quite a few core features are affected here. The change applies to the way variants and line items store and retrieve their names for unit values in the database. I think this is well covered in tests, but some things to look at would be:

- Create a new product and variant: does the name display it's units correctly in the shop? Like `Carrots 1 bunch`, or `Bottles (125ml)`
- Editing a product with variants and changing it's unit type, for example from `weight` to `volume`; do the product's variants get updated correctly as well?
- Viewing line items in places like reports, same thing. The names like `Bottles (125ml)` should be displayed correctly
- Using the "final weight/volume" feature in Bulk Order Management to change the final weight of line items.

#### Release notes

Changelog Category: Technical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
Note: the following database tables could all be dropped after this change: `spree_option_values`, `spree_option_types`, `spree_option_values_products`,  `spree_option_values_variants`, and `spree_option_values_line_items`.

File: spec/models/spree/line_item_spec.rb [hunk 1]
```diff
@@ -197,6 +197,14 @@ module Spree
         end
 
         it "includes soft-deleted products/variants" do
+          # Various erros on destroying a product, trying to destroy stock items for master,
+          # then finding they don't exist. Maybe the callbacks were doing some extra juju before...
+
+          pp v1
+          pp v1.stock_items
+          pp li3.quantity
+
```
Comment (mkllnk) [inline spec/models/spree/line_item_spec.rb:206]: Thank you for the explanation. Can we delete the debug output now? Maybe move the comment into the commit message and just explain what the reload is doing?
```suggestion
```

File: spec/models/spree/shipment_spec.rb [hunk 2]
```diff
@@ -57,7 +57,7 @@
     end
 
     context "variant was removed" do
-      before { variant.product.destroy }
+      before { variant.reload.product.destroy }
```
Comment (mkllnk) [inline spec/models/spree/shipment_spec.rb:60]: I still wonder why this is necessary. Is the variant or the product out of date?

**Review by mkllnk — APPROVED** [review]
I love this ratio:
Screenshot from 2023-05-08 10-16-10

This is a great reduction of complexity. Thank you. Excellent commits, well done. :sparkles:

File: spec/models/spree/line_item_spec.rb (hunk 1 above)
Comment (Matt-Yorkley) [reply spec/models/spree/line_item_spec.rb]: Oooooops, that wasn't meant to be commited :sweat_smile:  This was from a point when I was debugging issues with the variant factory setup that was causing lots of weird failures, but it's fixed now. :+1:

File: app/models/spree/line_item.rb [hunk 3]
```diff
@@ -42,12 +39,16 @@ class LineItem < ApplicationRecord
     before_save :update_inventory
     before_save :calculate_final_weight_volume, if: :quantity_changed?,
                                                 unless: :final_weight_volume_changed?
-    after_save :update_order
-    after_save :update_units
+    before_save :assign_units, if: ->(line_item) {
+      line_item.new_record? || line_item.changed_attributes.keys.include?("final_weight_volume")
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply app/models/spree/line_item.rb:43]: Ah I think there was more than one attribute here to start with but I cut the list down later :+1:

File: spec/models/spree/shipment_spec.rb (hunk 2 above)
Comment (Matt-Yorkley) [reply spec/models/spree/shipment_spec.rb:60]: It looks like it isn't needed. I'll drop it :+1:

**Review by mkllnk — APPROVED** [review]

**Comment by lin-d-hop** [comment]
Pausing on this for a sec awaiting:

- Better testing notes
- Shall we test this as a FF for product refactoring?

**Comment by Matt-Yorkley** [comment]
I added some test notes :+1:

**Review by jibees — APPROVED** [review]
Thanks for making our codebase cleaner... 🧼 

I have made a comment, and I have another one:
>  Also previous issues like Zapier or Matabase not being able to easily find what an item is called will be resolved, it'll be trivial to query those values in direct and simple SQL queries.

I'm a bit concerned about that: shouldn't we change anything that are related to these tools (I'd add N8N too)? Maybe some of SQL queries use the legacy behavior and therefor should be changed. What do you think?

Also, for the how to test section, I'd add some tests around the Bulk Order Management page, and specially when clicking a product in a line item in order to adjust total weight/volume... 

Nothing that should block the process, approving and move to Test Ready column.

**Review by mkllnk — COMMENTED** [review]
@jibees I think that you are raising some good questions there. Database integrations in N8N are a workaround and we never promised to keep the database the same. Those integrations need to be updated if they are affected by this. We should consider that in a release plan.

But I'm wondering which other parts of the application may break. The tests pass but that doesn't prove that they are using the new columns and that the old columns are unused. I think that we should start a new branch where we remove the old columns and run the specs on that branch to see what happens. My biggest concern is the move of `variant_unit` from the product to the variant. While that's a good move and it was planned anyway for Network 2.0, it suddenly allows variants of the same product to have different units and the app my not expect that. Or some part of the app may still write to the old column.

I'll move this to Dev Test to create another branch for running tests without the old columns. @Matt-Yorkley are you up for testing that?

**Comment by Matt-Yorkley** [comment]
Ah I see. Yeah okay, I'll try dropping all the associated tables. I avoided doing that in case it needed to be reverted in the short term but yes we can try it in a separate branch :+1:

**Comment by Matt-Yorkley** [comment]
Here it is https://github.com/openfoodfoundation/openfoodnetwork/pull/10854 :eyes: :heavy_check_mark: 

Five substantially sized tables dropped. Nice. https://github.com/openfoodfoundation/openfoodnetwork/pull/10854/commits/256d7b55e654ea97ac083c595574a7cec5a4ff92

**Comment by mkllnk** [comment]
Great, thank you! Glad it's green. Thorough work. I'll move this to Test Ready for some manual testing on staging.

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

Pinging to signal merge conflicts on this branch :-)

**Comment by filipefurtad0** [comment]
Sorry @Matt-Yorkley ,
I was just about to stage the PR - but noticed conflicts again.... I'll stage this asap after these are fixed again.

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

It's a major change, but could not spot any change in behavior - :tada: 

(1) I've created these products, before staging, and changed the unit_value option:
```

1 loaf of bread (ITEM)      ->       1 kg of bread
1kg of apples (weight)      ->       1 bag of apples
1L of milk (volume)         ->	     1 package of milk

1 sac of marbles (ITEM)
1g of cinnamon (weight)
1mL of milk (volume)
```

(2) After staging, I've changed these:

```
1 sac of marbles (ITEM)      ->       1 g of marbles (weight)
1g of cinnamon (weight)      ->       1 stick of cinammon
1mL of milk (volume)         ->	      10 L of milk
```

Checked the shopfront, order summary, order confirmation, emails... All as before.

Before staging, the summary page:

image

I've spotted a funny bug, neither introduced (nor fixed within this PR unfortunately :sweat_smile: ): notice that `milk` is displaying `0mL` (right), while we can clearly see that `unit_value` is `1`... I need to investigate this, but this is unrelated to these changes.

After staging:
image

And editing the variants:
image


And making the second set of changes (2)
image


Here `Pack By Product` report, vs shopfront:
image
 
The items are still grouped after name change; the same naming and units are displayed on both ends. 

This all looks fine to me. Which feels like a reason to celebrate :tada: 
Merging!

**Pull request merged**